### ISSUE TITLE
Docker image

### ISSUE THREAD

**Issue by Tchoupinax** [rstemmer/musicdb#28]
Hey !

It could be very nice if you can make a Docker image for your project, trying step just for testing your project and it takes a long time to install all dependancies, all python dependancies.. etc :)

Docker image are very fun to easily deploy but still more fun to simply launch and test in 10 seconds ! :)

Regards,

**Comment by rstemmer** [comment]
In the past I found lots of excuses to not dig into the deep of Docker. But you're right. For the purpose of trying out some hard-to-install software like MusicDB, Docker is perfect.

Your post motivates me to give it a try :)

Since I never worked with Docker it may take some time, but I'll try to spend the next weekend (07.09.19) on it.

**Comment by Tchoupinax** [comment]
Very happy to heard that ! :)

I can bring help about it. I began this afternoon but i stopped at Python dependancies steps, many dependancies cannot be found and i have very bad knowledges about Python environment.

Be free to ask help if needed. ;)

**Comment by rstemmer** [comment]
I created a branch called _docker_ with the current state of the setup for the Docker image.
The image will be based on fedora:latest.

Major task that is to do now is, to port and reduce the interactive install-script to a more precise setup for the image. This will take some time, but should not be a problem.

Furthermore I want to reduce the image size a bit by making the _id3edit_ tool optional for MusicDB. It is only needed for exporting music outside of the MusicDB environment, which is not a realistic use-case for a Test/Demo installation. Without _id3edit_ I do not have to install LLVM into the image.

**Comment by rstemmer** [comment]
I still have no idea for what use-case Docker is made, but it's barely suitable for providing a demo-image for software like MusicDB.

The main problem is, that I do not have a self-contained application. MusicDB requires an HTTP-Daemon, and Icecast running. The common work-around for this is to start everything from a script that them will be called by Docker's CMD instruction.
But then, only the Server runs, and there is no good way (only _run -it_) to add music by running other MusicDB command-line modules.

**Plan A**
I need to find a good trade-of between the needs of MusicDB and the capabilities of Docker.
A strategy can be to provide a very stripped installation without Icecast, without SSL encryption (which may change since MusicDB is not really capable of using non-encrypted websockets).
Furthermore the must be some demo-music be installed during the build-process of the image.
Without Icecast in the background, that will be "empty" mp3 files with only an album artwork.

**Plan B**
There is another idea I had while working with the Dockerfile. Why not generating a real virtual machine image that does not come with the limitations of Docker.
So instead of a Dockerfile I have a "GenerateImage.sh" script.
I may use a ready-to-use image like they are wide spread for Raspberry PIs. (Equivalent to _FROM_).
_COPY_ becomes _scp_, and _RUN_ becomes _ssh_.
Running such an image in _qemu_ also allows me _EXPOSEing_ all ports very easy.

**Comment by Tchoupinax** [comment]
You are the most qualified people to answer that question. :)

I bad understand what is the difficulty when you have started the server to "add music by running other MusicDB command-line modules". ?

What you do in "real machine" can be reproduced to Docker. Have only one process by container is a best practice and very useful for scalability, but for demonstration image it is not mandatory.

**Comment by rstemmer** [comment]
**I think it's done :)**

I successfully started a working container, started all servers, added music to MusicDB and accessed the WebUI.
The whole process of testing MusicDB should be quiet easy. Starting with the README, the user gets guided through each step. And they are all very simple.

Everything is still on the docker-branch.

What is missing in Docker some kind of boot process where all depending servers gets started (in my case, "httpd", "icecast" and "musicdb server"). This now has been solved by a script "conatiner-boot.sh".
After running this command, the interactive shell of the container can be used to execute other musicdb commands like "musicdb add" to add new music to MusicDB.

Now I only hope that nobody gets the stupid idea to use this Docker-Image in production.

Final steps to do:
* Write Documentation (mostly copying the README and run.sh tips into a reStructuredText file
* Rebuild the HTML-Documentation
* Updating the CHANGELOG and VERSION (minor bugfixes)
* Updating the Website, highlighting the easy-to-use demo that is now available :)

Plan B will be realized in future to generate a ready-to-run Raspberry Pi Image.
